### DISCUSSION TITLE
Why is there a maxMessageSize?

### DISCUSSION THREAD

**Discussion by caramboleyo** [geckosio/geckos.io#279] [Help / Questions]
I just tried to send tile data for a map and geckos says

> Error in sendMessage.ts:  maxMessageSize of 262144 exceeded

doesn't datachannel split it up in packages anyway?

btw my message is 211190bytes, does geckos send 50KB overhead per message?

**Comment by reececomo** [comment]
Side question for @caramboleyo - 200+KB of tile data seems **huge**

WebRTC Datachannels is probably not the right tool for that job. Another approach might be to download tile data as a .json file (or another file format), or streaming it in smaller chunks. Better yet, use something like websockets which can have a virtually unlimited message size.

I think the _practical_ maximum message size for wrtc datachannels is something like 16Kb (or maybe up to 64Kb in some browsers).

**Reply by reececomo** [reply]
For context, in the above reply I'm making an assumption here you have both a mixture of "realtime" data that you need to stream (via unreliable, UDP-like WebRTC Datachannels), and bulkier, required data for your map which needs to be 100% reliable. If you have any more context that might help.

**Reply by caramboleyo** [reply]
i am currently using websockets and read a lot of articles and discussions lately stating it is better to move on to webRTC and udp since its faster and less overhead.
currently when you pan the map all tiles get loaded in one request that contains the tiles appearing in your view. When units are in battle a lot of data comes in about their remaining values (hull,shields etc) and the hits they take.

**Reply by reececomo** [reply]
200kb of data would be a lot ("too much") for a single WebRTC message.

WebRTC data channels in geckos.io are "UDP-like" unreliable, unordered. Your map data probably needs to be sent reliably.

You could consider breaking up those into smaller sends ([and implement something like Glenn Fielders approach as inspiration](https://gafferongames.com/post/reliable_ordered_messages/)), but for this specific case it sounds like what you need is reliable, ordered messages that can take large chunks of data (200kb+). You might want to consider sticking with web sockets, or using both. It depends on how much of your game streams real-time. If you're currently sending raw uncompressed JSON, you may also want to see how much gain you can get compressing your packets into binary (geckosio provides [typed-array-buffer-schema](https://github.com/geckosio/typed-array-buffer-schema) which will do this for you, or use [typescript-binary](https://github.com/reececomo/typescript-binary)).

- You'll definitely want to read: [Sending Large Blocks of Data](https://gafferongames.com/post/sending_large_blocks_of_data/), and
- [UDP vs TCP](https://gafferongames.com/post/udp_vs_tcp/)

Feel free to share more specifics/screenshots or code samples from your game if you want more specific help. 😊
